dll/win32/msafd/misc/dllmain.c

```diff
@@ -2796,166 +2805,181 @@ WSPSetSockOpt(
+
+                case SO_RCVBUF:
+                    if (optlen < sizeof(ULONG))
+                    {
+                        if (lpErrno) *lpErrno = WSAEFAULT;
+                        return SOCKET_ERROR;
+                    }
+
+                    /* FIXME: We should not have to limit the packet receive buffer size like this. workaround for CORE-15804 */
+                    if (*(PULONG)optval > 0x2000)
+                        *(PULONG)optval = 0x2000;
+
+                    SetSocketInformation(Socket,
+                                         AFD_INFO_RECEIVE_WINDOW_SIZE,
+                                         NULL,
+                                         (PULONG)optval,
+                                         NULL,
+                                         NULL,
+                                         NULL);
+                    GetSocketInformation(Socket,
+                                         AFD_INFO_RECEIVE_WINDOW_SIZE,
+                                         NULL,
+                                         &Socket->SharedData->SizeOfRecvBuffer,
+                                         NULL,
+                                         NULL,
+                                         NULL);
+
+                    return NO_ERROR;
+
+                case SO_ERROR:
+                    if (optlen < sizeof(INT))
+                    {
+                        if (lpErrno) *lpErrno = WSAEFAULT;
+                        return SOCKET_ERROR;
+                    }
+
+                    RtlCopyMemory(&Socket->SharedData->SocketLastError,
+                                  optval,
+                                  sizeof(INT));
+                    return NO_ERROR;
+
+                case SO_SNDTIMEO:
+                    if (optlen < sizeof(DWORD))
+                    {
+                        if (lpErrno) *lpErrno = WSAEFAULT;
+                        return SOCKET_ERROR;
+                    }
+
+                    RtlCopyMemory(&Socket->SharedData->SendTimeout,
+                                  optval,
+                                  sizeof(DWORD));
+                    return NO_ERROR;
+
+                case SO_RCVTIMEO:
+                    if (optlen < sizeof(DWORD))
+                    {
+                        if (lpErrno) *lpErrno = WSAEFAULT;
+                        return SOCKET_ERROR;
+                    }
+
+                    RtlCopyMemory(&Socket->SharedData->RecvTimeout,
+                                  optval,
+                                  sizeof(DWORD));
+                    return NO_ERROR;
+
+                case SO_KEEPALIVE:
+                case SO_DONTROUTE:
+                    /* These go directly to the helper dll */
+                    goto SendToHelper;
+
+                default:
+                    /* Obviously this is a hack */
+                    ERR("MSAFD: Set SOL_SOCKET unknown optname %x\n", optname);
+                    return NO_ERROR;
+            }
```

Comment:
This bracket is the end of the `switch ()` statement from the `case SOL_SOCKET:`
Why isn't there any "break" here at the end? Is it because we have to fallback into these other cases that just `goto SendToHelper;` ? or ? (Note that the `WSPGetSockOpt` function does **NOT** make its `case SOL_SOCKET:` handling ultimately fall back to these cases.)

Reply:
Nice appreciation. So the case SOL_SOCKET in WSPSetSockOpt has an internal switch(optname), all the cases handled in switch(optname) ends returning something or gotoing, so in case SOL_SOCKET we shouldn't be falling to its next sibling case (line 2972). 

If you wish we can add a "return 0" or a "break;" to enhace readibility but Coverity will sign it as unreachable code because all the previous switch..case paths are closed. (If you see a not-closed path, please tell me)

Regarding WSPGetSockOpt, iiuc, SOL_SOCKET won't fall into IPPROTO_IP because the line 2574 returns 0.